java/vector/src/main/java/org/apache/arrow/vector/stream/MessageSerializer.java

```diff
@@ -120,6 +120,9 @@ public static ArrowBlock serialize(WriteChannel out, ArrowRecordBatch batch)
 
     long start = out.getCurrentPosition();
     int bodyLength = batch.computeBodyLength();
+    if (bodyLength % 8 != 0) {
+      bodyLength += 8 - (bodyLength % 8);
+    }
```

Comment:
You've stumbled on https://issues.apache.org/jira/browse/ARROW-412. Not padding the last buffer in Java is definitely a bug. 

Whether the padding to a 8-byte offset should be included in the Buffer metadata is something which should be discussed further and well-specified in the format documents.